File: Python_Solutions/Find_K_Pairs_with_Smallest_Sums.py
"""
Leetcode Challenge: Find K Pairs with Smallest Sums

You are given two integer arrays nums1 and nums2 sorted in 
non-decreasing order and an integer k.

Define a pair (u, v) which consists of one element from the first 
array and one element from the second array.

Return the k pairs (u1, v1), (u2, v2), ..., (uk, vk) with the smallest sums.

9/21/24
Mark Robuck
"""
import heapq
from typing import List

class Solution:
    """
    Find pairs of values from two lists in order of smallest sum
    Args:
        nums1(List[int]): list of sorted integers
        nums2(List[int]): list of sorted integers
        k(int): number of pairs to return
    Return:
        List[int]: pairs of integers from nums1, nums2 in order of smallest sum
    """
    def kSmallestPairs(self, nums1: List[int], nums2: List[int], k: int) -> List[List[int]]:
        #min-heap for maintaining the smallest sum pairs
        min_heap = []
        result = []
        
        #edge cases
        if not nums1 or not nums2 or k == 0:
            return result
        
        #add the first k pairs from nums1[0] and all of nums2
        for i in range(min(k, len(nums1))):
            heapq.heappush(min_heap, (nums1[i] + nums2[0], i, 0))
        
        #extract k smallest pairs
        while k > 0 and min_heap:
            current_sum, i, j = heapq.heappop(min_heap)
            result.append([nums1[i], nums2[j]])
            k -= 1
            
            #if possible, push the next pair (nums1[i], nums2[j+1]) into the heap
            if j + 1 < len(nums2):
                heapq.heappush(min_heap, (nums1[i] + nums2[j+1], i, j+1))
        
        return result
        

File: Java_Solutions/Find_K_Pairs_with_Smallest_Sums.java
/*
* Leetcode Challenge: Find K Pairs with Smallest Sums
*
* You are given two integer arrays nums1 and nums2 sorted in 
* non-decreasing order and an integer k.
*
* Define a pair (u, v) which consists of one element from the first 
* array and one element from the second array.
*
* Return the k pairs (u1, v1), (u2, v2), ..., (uk, vk) with the smallest sums.
*
* 9/21/24
* Mark Robuck
*/
import java.util.*;

class Solution {
    /*
    * Find pairs of values from two lists in order of smallest sum.
    * @params nums1, list of sorted integers
    * @params nums2, list of sorted integers
    * @params k, number of pairs to return
    * @return pairs of integers from nums1, nums2 in order of smallest sum
    */
    public List<List<Integer>> kSmallestPairs(int[] nums1, int[] nums2, int k) {
        //min-heap for maintaining the smallest sum pairs
        PriorityQueue<int[]> minHeap = new PriorityQueue<>((a, b) -> (a[0] + a[1]) - (b[0] + b[1]));
        List<List<Integer>> result = new ArrayList<>();
        
        //edge case
        if (nums1.length == 0 || nums2.length == 0 || k == 0) {
            return result;
        }
        
        //add the first k pairs from nums1[0] and all of nums2
        for (int i = 0; i < Math.min(k, nums1.length); i++) {
            minHeap.offer(new int[]{nums1[i], nums2[0], 0});
        }
        
        //extract k smallest pairs
        while (k > 0 && !minHeap.isEmpty()) {
            int[] current = minHeap.poll();
            result.add(Arrays.asList(current[0], current[1]));
            k--;
            
            //if possible, push the next pair (nums1[i], nums2[j+1]) into the heap
            if (current[2] + 1 < nums2.length) {
                minHeap.offer(new int[]{current[0], nums2[current[2] + 1], current[2] + 1});
            }
        }
        
        return result;
    }
}
